taichi/program.cpp

```diff
@@ -205,6 +218,10 @@ void Program::materialize_layout() {
       metal_runtime_ = std::make_unique<metal::MetalRuntime>(std::move(params));
     }
     TI_INFO("Metal root buffer size: {} B", metal_struct_compiled_->root_size);
+  } else if (config.arch == Arch::opengl) {
+    opengl::OpenglStructCompiler scomp;
+    opengl_struct_compiled_ = scomp.run(*snode_root);
```

Comment:
Maybe we can move `initialize_opengl()` here?

Reply:
Good idea! But does `materialize_layout` run every `ti.init()`? Or we should move `gl_inited` inside `initialize_opengl()`. My idea is to call `initialize_opengl()` before every function in `opengl_api.h` is called (done in later stages).

Reply:
yeah i think so, `materialize_layout` must run before any kernel execution, right? BTW, a few lines above, where you can see

```
metal_runtime_ = std::make_unique<metal::MetalRuntime>(std::move(params));
```

This does something similar to `initialize_opengl()`...

Reply:
I'm not very sure if they are similar... What `initialize_opengl()` mainly do is `glewInit()`, to fill GL APIs with function pointers instead of NULL. So it must be called before any other API is called.

Reply:
>  So it must be called before any other OpenGL API is called.

yeah, that's what i mean by similar.. `materialize_layout()` is a good place to put all such backend initialization procedures.. I believe it's guaranteed to be invoked before any kernel launches, therefore it's before any OpenGL API calls.